specs/core/0_fork-choice.md

```diff
@@ -130,13 +144,43 @@ def get_head(store: Store) -> Hash:
         head = max(children, key=lambda root: (get_latest_attesting_balance(store, root), root))
 ```
 
+#### `should_update_justified_checkpoint`
+
+```python
+def should_update_justified_checkpoint(store: Store, justified_checkpoint: Checkpoint) -> bool:
+    if get_current_slot(store) % SLOTS_PER_EPOCH < SAFE_SLOTS_TO_UPDATE_JUSTIFIED:
+        return True
+
+    justified_block = store.blocks[justified_checkpoint.root]
+    if justified_block.slot <= compute_start_slot_at_epoch(store.justified_checkpoint.epoch):
+        return False
+    if not get_ancestor(store, justified_checkpoint.root, store.blocks[justified_checkpoint.root].slot):
+        return False
+
+    return True
+```
+
 ### Handlers
 
 #### `on_tick`
 
 ```python
 def on_tick(store: Store, time: uint64) -> None:
+    previous_slot = get_current_slot(store)
+
+    # update store time
     store.time = time
+
+    current_slot = get_current_slot(store)
+    # Not a new epoch, return
+    if not (current_slot > previous_slot and current_slot % SLOTS_PER_EPOCH == 0):
+        return
+    # If new epoch and there are queued_justified_checkpoints, update if any is better than the best in store
+    if any(store.queued_justified_checkpoints):
+        best_justified_checkpoint = max(store.queued_justified_checkpoints, key=lambda checkpoint: checkpoint.epoch)
+        if best_justified_checkpoint.epoch > store.justified_checkpoint.epoch:
+            store.justified_checkpoint = best_justified_checkpoint
+    store.queued_justified_checkpoints = []
```

Comment:
See other comment. We can just track the `best_justified_checkpoint` instead of maintaining a whole queue?